File: solutions/julia/square-root/3/square-root.jl
function square_root(n)
    if n <= 0
        throw(DomainError(n))
    end
    left = 1
    right = n
    while left < right
        middle = (left + right + 1) ÷ 2
        
        if n ÷ middle >= middle
            left = middle
        else
            right = middle - 1
        end
    end
    left
end

File: solutions/swift/square-root/1/Sources/SquareRoot/SquareRoot.swift
struct SquareRoot {
   static func squareRoot(_ n: Int) throws -> Int {
      if n < 0 {
         throw(DomainError.Negative)
      }
      var left = Int(0)
      var right = n
      while left < right {
         let middle = (left + right + 1) / Int(2)
         if n / middle >= middle {
            left = middle
         } else {
            right = middle - 1
         }
      }
      return left
   }

   enum DomainError: Error {
      case Negative
   }
}
